File: internal/service/user/sync_status.go
package user

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/KazuhaHub/passwall-sub-panel/internal/domain"
)

// The task types a sync-status read covers. Fixed, not caller-supplied: the
// answer is only meaningful next to the set it was computed from, and widening
// the set changes the meaning of "no active tasks".
//
// This is deliberately NOT the same set HasPendingSync checks — that one omits
// user_migrate, so it can report "nothing pending" while a shared-client
// migration is still running.
var syncStatusTaskTypes = []domain.SyncTaskType{
	domain.SyncTaskUserDelete,
	domain.SyncTaskUserResync,
	domain.SyncTaskUserPushConfig,
	domain.SyncTaskUserMigrate,
}

const (
	// Response caps, not business limits. The extra row is only read to detect
	// truncation; it is never returned.
	syncStatusActiveLimit   = 20
	syncStatusTerminalLimit = 10
)

// SyncStatusState is what a read observed about local tasks. There is
// deliberately no value meaning "upstream is in sync": no active task is
// evidence about the queue, not about the panel. See ADR 0034.
type SyncStatusState string

const (
	SyncStatusActiveTasks   SyncStatusState = "active_tasks"
	SyncStatusNoActiveTasks SyncStatusState = "no_active_tasks"
)

// SyncStatusHistoryScope marks the terminal list as what is still stored, so a
// purged history is never read as "nothing ever ran".
const SyncStatusHistoryScopeRetained = "retained_only"

// SyncTaskView is the subset of a task a status read may expose. Payload,
// Summary and LastError are excluded: a task payload can carry subscription
// credentials, and an upstream error text can carry an endpoint or a token.
type SyncTaskView struct {
	ID         int64
	Type       domain.SyncTaskType
	Status     domain.SyncTaskStatus
	Attempts   int
	NextRunAt  time.Time
	CreatedAt  time.Time
	UpdatedAt  time.Time
	FinishedAt *time.Time
	// HasError reports that an error was recorded, not what it was. It does not
	// distinguish "still retrying" from "gave up" — Status does that.
	HasError bool
}

// SyncStatus is one observation of a user's sync tasks.
type SyncStatus struct {
	TargetType   string
	TargetID     int64
	TargetExists bool
	ObservedAt   time.Time

	CoveredTaskTypes []domain.SyncTaskType
	State            SyncStatusState

	ActiveTasks          []SyncTaskView
	ActiveTasksTruncated bool

	RecentTerminalTasks []SyncTaskView
	HistoryTruncated    bool
	HistoryScope        string
}

func syncTaskView(t *domain.SyncTask) SyncTaskView {
	return SyncTaskView{
		ID: t.ID, Type: t.Type, Status: t.Status,
		Attempts: t.Attempts, NextRunAt: t.NextRunAt,
		CreatedAt: t.CreatedAt, UpdatedAt: t.UpdatedAt, FinishedAt: t.FinishedAt,
		HasError: t.LastError != "",
	}
}

// SyncStatus reports what local sync work is observable for one user.
//
// It never fails because the target is missing: an admin looking at a deleted
// user still needs to see what happened to its retained tasks, so a missing row
// is reported as TargetExists=false rather than as an error. It DOES fail when
// the task store cannot answer — reporting that as "no active tasks" is the
// behaviour ADR 0034 exists to stop.
//
// The read touches no upstream panel and changes no task state.
func (s *Service) SyncStatus(ctx context.Context, userID int64) (*SyncStatus, error) {
	if s.tasks == nil {
		return nil, fmt.Errorf("sync status: task store not configured: %w", domain.ErrUnavailable)
	}

	out := &SyncStatus{
		TargetType:       "user",
		TargetID:         userID,
		ObservedAt:       time.Now().UTC(),
		CoveredTaskTypes: append([]domain.SyncTaskType(nil), syncStatusTaskTypes...),
		HistoryScope:     SyncStatusHistoryScopeRetained,
	}

	switch _, err := s.users.GetByID(ctx, userID); {
	case err == nil:
		out.TargetExists = true
	case errors.Is(err, domain.ErrNotFound):
		out.TargetExists = false
	default:
		return nil, fmt.Errorf("sync status: resolve target: %w", domain.ErrUnavailable)
	}

	// One extra row per query is how truncation is detected without a count.
	active, err := s.tasks.ListActiveByTarget(ctx, syncStatusTaskTypes, "user", userID, syncStatusActiveLimit+1)
	if err != nil {
		return nil, fmt.Errorf("sync status: list active tasks: %w", domain.ErrUnavailable)
	}
	if len(active) > syncStatusActiveLimit {
		out.ActiveTasksTruncated = true
		active = active[:syncStatusActiveLimit]
	}
	out.ActiveTasks = make([]SyncTaskView, 0, len(active))
	for _, t := range active {
		out.ActiveTasks = append(out.ActiveTasks, syncTaskView(t))
	}

	terminal, err := s.tasks.ListTerminalByTarget(ctx, syncStatusTaskTypes, "user", userID, syncStatusTerminalLimit+1)
	if err != nil {
		return nil, fmt.Errorf("sync status: list terminal tasks: %w", domain.ErrUnavailable)
	}
	if len(terminal) > syncStatusTerminalLimit {
		out.HistoryTruncated = true
		terminal = terminal[:syncStatusTerminalLimit]
	}
	out.RecentTerminalTasks = make([]SyncTaskView, 0, len(terminal))
	for _, t := range terminal {
		out.RecentTerminalTasks = append(out.RecentTerminalTasks, syncTaskView(t))
	}

	if len(out.ActiveTasks) > 0 {
		out.State = SyncStatusActiveTasks
	} else {
		out.State = SyncStatusNoActiveTasks
	}
	return out, nil
}

File: web-react/src/api/syncStatus.ts
import { client } from './client'
import type { ReadOptions } from './requestOptions'

/**
 * What a status read observed about LOCAL sync work. There is deliberately no
 * value meaning "upstream is in sync": no active task is evidence about the
 * queue, not about whether a panel applied the configuration. See ADR 0034.
 */
export type SyncStatusState = 'active_tasks' | 'no_active_tasks'

/** The subset of a task the API exposes. Payload, summary and the raw upstream
 *  error are not sent — a payload can carry subscription credentials. */
export interface SyncTaskView {
  id: number
  type: string
  status: string
  attempts: number
  next_run_at: string
  created_at: string
  updated_at: string
  finished_at?: string
  /** Only that an error was recorded; never the text. */
  has_error: boolean
}

export interface SyncStatus {
  target_type: string
  target_id: number
  target_exists: boolean
  /** When the server read its own task store. Says nothing about the panel. */
  observed_at: string
  covered_task_types: string[]
  state: SyncStatusState
  active_tasks: SyncTaskView[]
  /** The lists are capped, so a count read off them is a floor, not a total. */
  active_tasks_truncated: boolean
  recent_terminal_tasks: SyncTaskView[]
  history_truncated: boolean
  /** Always 'retained_only': purged rows are absent, and absence must not be
   *  read as "never ran". */
  history_scope: string
}

/** The stable code the server returns when it could not answer. */
export const SYNC_STATUS_UNAVAILABLE_CODE = 'sync_status_unavailable'

export async function getSyncStatus(userId: number, opts: ReadOptions = {}): Promise<SyncStatus> {
  const { data } = await client.get<SyncStatus>(`/admin/users/${userId}/sync-status`, {
    signal: opts.signal,
  })
  return data
}

/** The self-service read: the target is the session, never a parameter. */
export async function getMySyncStatus(opts: ReadOptions = {}): Promise<SyncStatus> {
  const { data } = await client.get<SyncStatus>('/user/me/sync-status', { signal: opts.signal })
  return data
}
